### PR TITLE
feat: use event listener hook and refactor usages

### PR DESCRIPTION
**Pull request by capJavert**
### Describe what this PR does
- Introduce new `useEventListener` hook that supports all different event handlers and iterations of `addEventListener`
- Refactor all usages of `useWindowEvents` to `useEventListener`
- Move `validateKey` functionality to handlers for better visibility vs hiding it in hook implementation (also they are used in less places
- Available on [preview](https://preview.app.daily.dev/)

TODO:
- [x] end to end test places where hooks were used to assure everything is working with `data.eventKey` handling
- [x] tested social login/registration
- [x] tested email login and registration
- [x] tested push notification enable
- [x] tested interactive popup close from outside click (global click handler)
- [x] tested companion login data refetch
- [x] tested popstate reload 
- [x] tested logout 

AS-25 #done

### CODE REVIEW

File: packages/shared/src/hooks/useEventListener.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+export type TUseEventListenerOptions =
+  | boolean
+  | AddEventListenerOptions
+  | undefined;
+
+export interface MessageEventData {
+  eventKey?: string;
+}
+
+const useEventListener = <
```
Comment (capJavert) [inline packages/shared/src/hooks/useEventListener.ts:95]: Whole hook and tests are mostly the same logic from [here](https://github.com/kickassCoderz/kickass-toolkit/blob/16d2b6264c7dd6b09b2412cc41ea9df0b7815f49/packages/react/src/hooks/useEventListener/useEventListener.ts), could not use the package since we are still on react@17.x so copied everything for now.

File: packages/shared/src/components/auth/AuthOptions.tsx [hunk 2]
```diff
@@ -239,70 +238,67 @@ function AuthOptions({
...
-  useWindowEvents<SocialRegistrationFlow>(
-    'message',
-    AuthEvent.SocialRegistration,
-    async (e) => {
-      if (e.data?.flow) {
-        const connected = await getKratosFlow(
-          AuthFlow.Registration,
-          e.data.flow,
-        );
+  useEventListener(globalThis, 'message', async (e) => {
```
Comment (rebelchris) [inline packages/shared/src/components/auth/AuthOptions.tsx:241]: This one will require some manual/preview verification before we launch it.
Mostly on both email registration and social, just to exclude edge-cases.

File: packages/shared/src/hooks/useEventListener.spec.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    expect(removeSpy).toHaveBeenCalledTimes(0);
+
+    rerender();
+
+    expect(addSpy).toHaveBeenCalledTimes(1);
+    expect(removeSpy).toHaveBeenCalledTimes(0);
+
+    unmount();
+
+    expect(addSpy).toHaveBeenCalledTimes(1);
```
Comment (rebelchris) [inline packages/shared/src/hooks/useEventListener.spec.ts:68]: should this be called on unmount?

File: packages/webapp/pages/posts/[id]/index.tsx [hunk 4]
```diff
@@ -95,14 +92,10 @@ const PostPage = ({ id, initialData }: Props): ReactElement => {
...
-  useWindowEvents(
-    'popstate',
-    CHECK_POPSTATE,
-    // @NOTE see https://dailydotdev.atlassian.net/l/cp/dK9h1zoM
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    useCallback(() => router.reload(), []),
-    { validateKey: false },
-  );
+
+  useEventListener(globalThis, 'popstate', () => {
```
Comment (rebelchris) [inline packages/webapp/pages/posts/[id]/index.tsx:96]: what did we use the validateKey and callback for again?

**Review by rebelchris — APPROVED** [review]
Nothing to add as hook existed on your side.
we definitely need to test extensively and the critical flows before/after merging

File: packages/shared/src/components/auth/AuthOptions.tsx [hunk 5]
```diff
@@ -239,70 +238,67 @@ function AuthOptions({
...
-      if (!bootResponse.data.user || !('email' in bootResponse.data.user)) {
-        trackEvent({
-          event_name: AuthEventNames.SubmitSignUpFormError,
-          extra: JSON.stringify({
-            error: 'Could not find email on social registration',
-          }),
-        });
-        return displayToast(labels.auth.error.generic);
+      if (
+        [4010002, 4010003, 4000007].includes(connected?.ui?.messages?.[0]?.id)
```
Comment (kopancek) [inline packages/shared/src/components/auth/AuthOptions.tsx:261]: What do these magic numbers mean? Should we add const variable with more meaningful names?

File: packages/shared/src/components/auth/AuthOptions.tsx [hunk 6]
```diff
@@ -239,70 +238,67 @@ function AuthOptions({
...
-    'message',
-    AuthEvent.SocialRegistration,
-    async (e) => {
-      if (e.data?.flow) {
-        const connected = await getKratosFlow(
-          AuthFlow.Registration,
-          e.data.flow,
-        );
+  useEventListener(globalThis, 'message', async (e) => {
+    if (e.data?.eventKey !== AuthEvent.SocialRegistration) {
```
Comment (kopancek) [inline packages/shared/src/components/auth/AuthOptions.tsx:242]: Can this condition be added directly to the `useEventListener`? E.g. by passing another parameter to the `useEventListener` hook. I see that we have the same behavior everywhere where the hook is used (returning undefined), just the condition is different...

File: packages/shared/src/hooks/useAuthErrors.ts [hunk 7]
```diff
@@ -1,10 +1,15 @@
...
 export function useAuthErrors(): void {
   const { displayToast } = useToastNotification();
-  useWindowEvents<ErrorEvent>('message', AuthEvent.Error, async (e) => {
+
+  useEventListener(globalThis, 'message', async (e) => {
+    if (e.data?.eventKey !== AuthEvent.Error) {
+      return;
+    }
+
     if (!e.data?.id) {
```
Comment (kopancek) [inline packages/shared/src/hooks/useAuthErrors.ts:13]: Since the return value is the same, these conditions can be merged, e.g. 
```suggestion
    if (e.data?.eventKey !== AuthEvent.Error || !e.data?.id) {
```

**Review by kopancek — COMMENTED** [review]
I don't really have context here. Looks good to me, just spotted some potential places for improving the code or making it easier to understand.

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 5 above)
Comment (sshanzel) [reply packages/shared/src/components/auth/AuthOptions.tsx:261]: These are error codes from Kratos. We should have an enum to hold these values `KRATOS_ERROR`.

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 6 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx:242]: This was what the old hook did and was hidden with `validateKey`. I think this (while adding some code) at least keeps the condition visible when looking at the related code. And we don't really use it on much of the places.

File: packages/shared/src/hooks/useAuthErrors.ts (hunk 7 above)
Comment (capJavert) [reply packages/shared/src/hooks/useAuthErrors.ts:13]: Would it not be less readable then?

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 5 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx:261]: Yeah I could add the enum here 👌 @sshanzel do you know from where are those codes taken/documented?

File: packages/webapp/pages/posts/[id]/index.tsx (hunk 4 above)
Comment (capJavert) [reply packages/webapp/pages/posts/[id]/index.tsx:96]: Here it was not used because `validateKey: false`, on other places I moved checks to the callbacks. From what I seen the kratos `message` events had `data.eventKey` to differentiate so it was a guard to execute logic only for specific event.

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 6 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx:242]: I could add the new hook like `useKratosMessageListener` that would encapsulate that logic and subscribe to specific key?

File: packages/shared/src/hooks/useEventListener.spec.ts (hunk 3 above)
Comment (capJavert) [reply packages/shared/src/hooks/useEventListener.spec.ts:68]: It checks that listener was not added after unmount eg. it was still called only once during mount.

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 2 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx:241]: Yes!

File: packages/shared/src/hooks/useEventListener.spec.ts (hunk 3 above)
Comment (rebelchris) [reply packages/shared/src/hooks/useEventListener.spec.ts:68]: I see :) perfect in that case

File: packages/webapp/pages/posts/[id]/index.tsx (hunk 4 above)
Comment (rebelchris) [reply packages/webapp/pages/posts/[id]/index.tsx:96]: I see, ok no worries in that case.

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/components/auth/AuthOptions.tsx:241]: Will this be attached to the window with just this?

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 5 above)
Comment (sshanzel) [reply packages/shared/src/components/auth/AuthOptions.tsx:261]: Should be available through this document: sendAudioData

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 2 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx:241]: Yes, and using `globalThis` nomenclature instead of window as we do recently.

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 5 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx]: Added in 52ab383713c09115c39c2380ed2304d8ba290c11

File: packages/shared/src/components/auth/AuthOptions.tsx (hunk 2 above)
Comment (capJavert) [reply packages/shared/src/components/auth/AuthOptions.tsx:242]: I tested the cases above (updated description).

File: packages/shared/src/hooks/useAuthErrors.ts (hunk 7 above)
Comment (kopancek) [reply packages/shared/src/hooks/useAuthErrors.ts:13]: I don't have a strong preference. To me the readability is the same. Having 2 conditions below each other when it can be defined with OR operator, I always prefer the OR operator.

If readability is a concern, you can split the condition into two lines, e.g.:
```suggestion
    if (e.data?.eventKey !== AuthEvent.Error ||
        !e.data?.id)
      return;
    }
```

Comment (kopancek) [reply packages/shared/src/hooks/useAuthErrors.ts:13]: Also, do not feel obliged. This is not blocking the merge from my point of view, just to be clear :)

Comment (capJavert) [reply packages/shared/src/hooks/useAuthErrors.ts:13]: Above would not be allowed by our linter but basically not about code readability, its about logic readability. 

Multiple `||` and `&&` etc. quickly become hard to parse logically when looking the code. Throw in a negation `!` and you are looking at something that becomes a monster really fast. 

Early return pattern like above solves it in 99% of the time while keeping conditions nice and lean. Would love to talk about it more at some other time 🙏
